pilot/pkg/status/state.go

```diff
@@ -103,7 +103,11 @@ func (c *DistributionController) Start(stop <-chan struct{}) {
 
 	// this will list all existing configmaps, as well as updates, right?
 	ctx := NewIstioContext(stop)
-	go c.cmInformer.Run(ctx.Done())
+	// Make sure informaer and DistributionController use context separately due to
+	// panic when losing and regaining leader election.
```

Comment:
I am a bit confused why this occurs. The code seems to make sense before... clearly it wasn't but I don't understand *why* not. For each leader election cycle, we pass in a `stop`. This is constructed from a context from the leader election library. Each election cycle will get a new one; the context.Done channel cannot be closed.

So which channel are we closing twice?

Reply:
@howardjohn According to the logs you mentioned in [issue#33095](https://github.com/istio/istio/issues/33095), it's obvious that stop channel is closed twice and still sent data after its close.  
I notice that there are stop channels for analysis controller and status controller,  the former use the channel to control the background analysis process [here](https://github.com/istio/istio/blob/c7e9ef8197a3fd948ce379a5513d2624149271a9/pilot/pkg/bootstrap/configcontroller.go#L276), and the latter shares the channel to its controller's informer to do watching and I believe that such sharing caused this issue.  Based on above investigation, I make a informer channel to status controller's informer, and the old `stop` channel can also stop the informer channel to avoid such twice channel close and still send data after its close.

Reply:
Please refer to [here ](https://github.com/kubernetes/kubernetes/issues/59822) for helping thinking on our informer issue, the issue were closed but it seems that concurrently calling of informer Run() function may still cause panic. However, make a new informer stop channel can handle this issue and the informer channel can also be controlled by old channel.

Reply:
From the logs information,  it seems that there are concurrently calling of informer Run() function may cause such panic, so I think this panic should be reproduce, @howardjohn is that possible for now?